File: CS2336/Lab03/Java/Solution/DoubleArrayBag.java
// File: DoubleArrayBag.java 

import java.util.Arrays;
import java.util.Objects;

/******************************************************************************
* An DoubleArrayBag is an unordered collection of double numbers and in which
* the same number may appear multiple times -- the bag's capacity can grow as
* needed and can be reduced.
*
* @note.
*   Because of the slow linear algorithms of this
*   class, large bags will have poor performance.
*
* @author
*    H. Paul Haiduk with credit given to Michael Main
*
* @version
*    September, 2020
*
******************************************************************************/
public class DoubleArrayBag implements Cloneable  {
   // Invariant of the DoubleArrayBag class:
   //   1. The number of elements in the bag is in the instance variable 
   //      used, which is no more than data.length.
   //   2. For an empty bag, we do not care what is stored in any of data;
   //      for a non-empty bag, the elements in the bag are stored in data[0]
   //      through data[used-1], and we don't care what's in the
   //      rest of data.
   //   3. Capacity of this bag at any given time is this.data.length

   private double[ ] data;
   private int used; 
   
   /**
   * Initialize an empty bag
   * @post.
   *   This bag is empty and has a capacity of 1.
   * @exception OutOfMemoryError
   *   Indicates insufficient memory for allocating the array 
   **/   
   public DoubleArrayBag( ) {
      this.used = 0;
      this.data = new double[1];
   }

   /**
    * Initialize an empty bag having a capacity of initialCapacity
    * @param
    *   initialCapacity an int greater than 0
    * @pre.
    *   initialCapacity must be greater than 0
    * @post.
    *   This bag is empty and has a capacity of initialCapacity
    * @exception
    *   IllegalArgumentException thrown if initialCapacity not greater than 0
    *   OutOfMemoryError thrown if insufficient memory for allocating the array
    **/
   public DoubleArrayBag( int initialCapacity ) {
       if ( initialCapacity < 1 )
           throw new IllegalArgumentException("initialCapacity < 1");
       this.used = 0;
       this.data = new double[initialCapacity];
   }

   /**
    * Initialize a new bag as an exact copy of source
    * @param source
    *   reference to bag that is to be copied
    * @exception NullPointerException
    *   occurs when source is null
    * @post.
    *   This bag is a copy of source and has a capacity
    *   of number of elements in source
    **/
   public DoubleArrayBag( DoubleArrayBag source ) {
       if ( source == null ) {
           throw new NullPointerException("source must not be null");
       }

       this.used = source.used;
       int myCapacity = source.used;
       if (myCapacity < 1) myCapacity = 1;
       this.data = new double[myCapacity]; //NOTE: only allocating needed space
       /* below logic we would write if we did not have System.arraycopy
       for (int index = 0; index < source.used; ++index) {
           this.data[index] = source.data[index];
       }
       */
       /* Documentation from Oracle about System.arraycopy is:
        * arraycopy(Object src, int srcPos, Object dest, int destPos, int length)
        * Copies an array from the specified source array, beginning at the specified
        * position, to the specified position of the destination array.
        **/
       System.arraycopy(source.data, 0, this.data, 0, source.used);
   }

     
   /**
   * Generate a copy of this bag.
   * @post.
   *    x.clone() != x
   *    x.clone().getClass() == x.getClass()
   *    x.clone().equals( x )
   *
   * @return
   *   The return value is a copy of this bag. Subsequent changes to the
   *   copy will not affect the original, nor vice versa.
   *
   **/ 
   public DoubleArrayBag clone( ) {  // Clone an DoubleArrayBag object.
      DoubleArrayBag answer;
      
      try {
         answer = (DoubleArrayBag) super.clone( );
         answer.data = data.clone( );
         return answer;
      }
      catch (CloneNotSupportedException e) {  
         // This exception should not occur. But if it does, it would probably
         // indicate a programming error that made super.clone unavailable.
         // The most common error would be forgetting the "Implements Cloneable"
         // clause at the start of this class.
         throw new RuntimeException ("This class does not implement Cloneable");
      }
   }

      
   /**
    * Compare this DoubleArrayBag to another object for equality of value
    * @param other
    *   reference to another DoubleArrayBag
    *
    * @post.
    *   x.equals(x) is true
    *   if x.equals(y) then y.equals(x)
    *   x.equals(null) is false
    *
    * @return
    *   true if number of elements in this and other are the same AND if
    *   the values of all the elements in the bag are the same and in the
    *   same position in the bag
    *
    * @note.
    *   If the value of other is null, then the return is false
    **/
   @Override
   public boolean equals(Object other) {
       if ( other == null ) return false;

       boolean isEqual = false;
       if ( other instanceof DoubleArrayBag ) {
           DoubleArrayBag candidate = (DoubleArrayBag) other;
           //are we comparing this with this
           if (this == other) return true;

           if ( candidate.used != this.used ) return false;
           if ( candidate.hashCode() != this.hashCode() ) return false;

           int index = 0;
           isEqual = true;
           while ( isEqual && index < this.used ) {
               if ( this.data[ index ] != candidate.data[ index ] )
                   isEqual = false;
               else
                   index++;
           }
       }
       return isEqual;
   }

   /**
    * Potentially increase capacity of this bag
    *
    * @pre.
    *   newCapacity must be greater than 0
    *
    * @post.
    *   The bag's capacity is at least newCapacity.  If the capacity
    *   was already at or greater than newCapacity, then the capacity
    *   is left unchanged.
    *
    * @param
    *   newCapacity an integer greater than 0
    *
    * @exception
    *   OutOfMemoryError if not enough dynamic available to allocate
    *   additional space
    * @exception
    *   IllegalArgumentException if newCapacity not greater than 0
    *
    **/
   public void ensureCapacity( int newCapacity ) {
       if ( newCapacity < 1)
           throw new IllegalArgumentException("newCapacity < 1");

       if ( this.data.length < newCapacity ) {
           //STUDENT WORK GOES HERE

           double[] newData = new double[newCapacity];

           for (int i = 0; i < used; i++)
               newData[i] = data[i];

           data = newData;
       }
   }

   /**
    * returns the current capacity of the bag
    * @post.
    *   This method does not alter state of the bag
    * @return
    *   an int that represents total capacity of this bag
    **/
   public int getCapacity( ) {
       return this.data.length;
   }


   /**
    * @return the hash code for this object
    *
    */
   @Override 
   public int hashCode( ) {
       int usedHashValue = Integer.hashCode( this.used );
       int dataHashValue = 0;

       for ( int i = 0; i < used; ++i ) {
           dataHashValue += Double.hashCode( this.data[i] );
       }

       return usedHashValue + dataHashValue;

   }

   /**
   * Erases all copies of a specified element from this bag if target exists in bag.
   * @param target
   *   the element(s) to remove from the bag
   * @post.
   *   If target was found in the bag, then all copies of
   *   target have been removed and the method returns number of items removed. 
   * @return
   *   int value from 0 to number of items erased from bag
   **/
   public int erase( double target ) {
       int count = 0;
       int index = 0;

       while ( index < used ) {
           boolean found = false;
           while ( !found && index < this.used ) {
               if ( target == this.data[ index ] )
                   found = true;
               else
                   ++index;
           }
           if (found) {
               this.data[ index ] = this.data[ --this.used ];
               ++count;
           }
       }

       return count;
   }
              
   /**
   * Remove one copy of a specified element from this bag.
   * @param target
   *   the element to remove from the bag
   * @post.
   *   If target was found in the bag, then one copy of
   *   target has been removed and the method returns true. 
   *   Otherwise the bag remains unchanged and the method returns false. 
   * @return
   *   true or false depending on whether target exists in the bag
   **/
   public boolean erase_one(double target) {
       int index = 0; // The location of target in the data array.
       boolean found = false;

       while (!found && index < this.used) {
           if ( target == this.data[index] )
               found = true;
           else
               ++index;
       }
       
       if (found) 
           this.data[index] = this.data[--this.used];

       return found;
   }
   
   /**
   * Add a new element to this bag doubling capacity if needed
   * 
   * @param newItem
   *   the new element that is being inserted
   * @post.
   *   A new copy of the element has been added to this bag.
   **/
   public void insert(double newItem) {
       if ( used == this.data.length ) {
           this.ensureCapacity( this.data.length * 2 );
       }
       this.data[ used++ ] = newItem;
   }

   /**
   * Add the contents of another bag to this bag.
   * @param addend
   *   a bag whose contents will be added to this bag
   * @pre.
   *   The parameter, addend, is not null. 
   * @post.
   *   The elements from addend have been added to this bag.
   * @exception NullPointerException
   *   Indicates that addend is null. 
   * @exception OutOfMemoryError
   *   Indicates insufficient memory to add addend to this bag
   **/
   public void plusEquals(DoubleArrayBag addend) {
       if ( addend == null ) {
           throw new NullPointerException("Cannot add from null object");
       }

       //STUDENT WORK GOES HERE

       if (addend.used > 0) {
           ensureCapacity(used + addend.used);

           int amt = addend.used;
           for (int i = 0; i < amt; i++)
               data[used++] = addend.data[i];
       }
   }   

   /**
   * Determine the number of elements in this bag.
   * @return
   *   the number of elements in this bag
   * @post.
   *   the bag is not altered by this method
   **/ 
   public int size( ) {
       return used;
   }
   
   /**
   * Accessor method to count the number of occurrences of a particular element
   * in this bag.
   * @param target
   *    the element for which number of occurrences will be counted 
   * @post.
   *    the bag is not altered by this method
   * @return
   *    the number of times that target occurs in this bag
   **/
   public int occurrences(double target) {
       int answer = 0;
      
       for (int index = 0; index < this.used; ++index) {
           if (target == this.data[index]) ++answer; 
       }
      return answer;
   }

   /**
    * This method renders the bag's contents into a human readable form
    * @return
    *    a string representation of the bag
    * @post.
    *    The bag is not altered by this method
    **/
   @Override
   public String toString() {
       // StringBuffer is used since its contents are mutable.  Strings are
       // immutable and, thus, very inefficient for such purpose
       StringBuffer sb = new StringBuffer();
       sb.append("DoubleArrayBag with ");
       sb.append(this.size());
       sb.append(" elements: [");
       if ( this.size() == 0 ) {
           sb.append(" ]");
           sb.append(" Capacity: "); 
           sb.append(this.getCapacity());
           return sb.toString();
       }
       for (int i=0; i < this.size()-1; ++i) {
           sb.append(" ");
           sb.append(this.data[i]);
           sb.append(",");
       }
       sb.append(" "); sb.append(this.data[this.size()-1]);
       sb.append(" ]");
       sb.append(" Capacity: "); sb.append(this.getCapacity());
       return sb.toString();
   }

   /**
    * reduce capacity of this bag to current size if there is
    * excess capacity
    *
    * @post.
    *   capacity of this bag is reduced to the current number
    *   of items in bag or left unchanged if capacity equals to
    *   number of items in bag but must be at least 1
    *
    * @exception
    *   OutOfMemoryError if not enough dynamic memory to
    *   support allocation and deallocation of memory
    **/
    public void trimToSize( ) {
        if ( this.used < this.data.length ) {
            int newCapacity;
            if ( this.used <= 1 )
                newCapacity = 1;
            else
                newCapacity = this.used;

            // STUDENT WORK GOES HERE

            double[] newData = new double[newCapacity];

            for (int i = 0; i < used; i++)
                newData[i] = data[i];

            data = newData;
        }
    }

   /**
   * Create a new bag that contains all the elements from two other bags -- note that
   * this is a class method NOT an instance method and must be called with class
   * name qualifier.
   * @param b1
   *   the first of two bags
   * @param b2
   *   the second of two bags
   * @pre.
   *   Neither b1 nor b2 is null, and b1.size() + b2.size() cannot
   *   exceed MAX_CAPACITY
   * @post.
   *   bag referenced by b1 and bag referenced by b2 are not altered
   * @return
   *   the union of b1 and b2 as a new DoubleArrayBag
   * @exception NullPointerException
   *   Indicates that one of the arguments is null.
   * @exception OutOfMemoryError
   *   Indicates insufficient memory for the new bag.  Only happens
   *   when not enough dynamic memory to allocate a new bag of
   *   adequate size.
   **/   
   public static DoubleArrayBag union(DoubleArrayBag b1, DoubleArrayBag b2) {
      if ( b1 == null || b2 == null ) {
          throw new NullPointerException("one or both bags reference is null");
      }

      // guarantee that the new bag has exactly the capacity needed to
      // hold all items from both b1 and b2 and then copy the items from
      // b1 and then b2 into the new bag

      //DoubleArrayBag newBag = null; // = new DoubleArrayBag( ?? );
      //STUDENT WORK HERE
      
      DoubleArrayBag newBag = b1.clone();

      if (b2.used > 0) {
          newBag.ensureCapacity(newBag.used + b2.used);

          for (int i = 0; i < b2.used; i++)
              newBag.data[newBag.used++] = b2.data[i];
      }

      return newBag;
   }

}
           

File: CS2336/Lab03/Python/Solution/DoubleArrayBag.py
#!/usr/bin/env python3

import array    # provides access to primitive arrays
from typing import TypeVar

DoubleArrayBag = TypeVar('DoubleArrayBag')
Number = TypeVar('Number', int, float)


class DoubleArrayBag(object):
    '''
    An DoubleArrayBag is a collection of double numbers

    Limitations:
        1.  The capacity of an DoubleArrayBag can grow to an arbitrarily
            large number of items after the initial bag creation.  This
            growth is limited by the amount of available freee dynamic
            memory allowed by Python.
        2.  Because of the linear algorithms used in this implementation.
            large DoubleArrayBags will have poor performance.
        3.  For efficiency, one should create an DoubleArrayBag with a
            capacity near what the user expects. This avoids unnecessary
            dynamic memory allocation.
    '''
#   Class Invariant:
#       1.  The number of elements in the bag is stored in the
#           instance variable __used which can be no greater than
#           len(__data)
#       2.  For an empty bag, we do not care what is stored in
#           any element of __data.  For a non-empty bag, the elements
#           are stored at __data[0] through __data[__used-1]
#           and we don't care what is stored in any elements beyond
#           that.
#       3.  The array __data can store only double items.  Attempts
#           to store items of other types will cause a run-time error
#           to be raised.
#       4.  Capacity of this bag at any given time is len(self.__data)

    def __init__(self, initialCapacity: int = 1) -> None:
        '''
        Precondition:   initialCapacity must be > 0

        Postcondition:  The bag has been initialized
                        and is in a well-defined, empty
                        state with initialCapacity elements for
                        storage

        Raises:         MemoryError if not able to allocate space to
                        accommodate initialCapacity items
        '''
        if initialCapacity is None or type(initialCapacity) != int:
            raise TypeError("initialCapacity must be an int non-null")
        if initialCapacity < 1:
            raise ValueError("initialCapacity must be > 0")

        self.__used = 0
        self.__data = array.array('d', range(0, initialCapacity))

    # end __init__

    def copy(self) -> DoubleArrayBag:
        '''
        this behaves like a copy constructor that makes a detailed copy of self

        NOTE: the capacity of the copy matches the elements in self unless
              the bag is empty, in which case it should be at least 1

        Postcondition:
            x.copy() is x returns false
            type(x.copy()) == type(x)
            x.copy() == x

            internal capacity of copy is just enough to hold self.__used
            items

        Returns:    a new DoubleArrayBag object that contains a copy of all
                    the elements in self and its capacity is equal to
                    number of elements in this bag

        Raises:     MemoryError if dynamic memory allocation fails
        '''
        copyCapacity = self.__used
        if (copyCapacity < 1):
            copyCapacity = 1
        # end if

        newBag = DoubleArrayBag(copyCapacity)

        # STUDENT IMPLEMENTATION GOES HERE
        #   must copy all elements from self.__data into
        #   the newBag and update any attributes to
        #   adhere to the class invariant

        for i in range(self.__used):
            newBag.__data[i] = self.__data[i]
        # end for

        newBag.__used = self.__used

        return newBag
    # end copy

    def size(self) -> int:
        '''
        Returns:  an integer representing numbers of items in the bag.
        '''
        return self.__used
    # end size

    def __eq__(self, other: DoubleArrayBag) -> bool:
        '''
        compare self with an other DoubleArrayBag

        Precondition:
            other must be an instance of DoubleArrayBag

        Postconditions:
            x == x is True
            if x == y then y == x
            x != None is True

        Args:
            other must be an instance of DoubleArrayBag

        Returns:
            boolean whether self is equal to other in number of items
            in the bag and in the location and value of the items in the bag

        NOTE:  if the value of other is None, return False
        '''
        if other is None:
            return False
        # end if

        if not isinstance(other, self.__class__):
            return False
        # end if

        if self is other:
            return True
        # end if

        if self.__used != other.__used:
            return False
        # end if

        index = 0
        isEqual = True  # assume equality
        while isEqual and index < self.__used:
            if self.__data[index] != other.__data[index]:
                isEqual = False
            else:
                index += 1
        # end while
        return isEqual
    # end __eq__

    def __ne__(self, other: DoubleArrayBag) -> bool:
        return not self == other
    # end __ne__

    def ensureCapacity(self, newCapacity: int) -> None:
        '''
        potentially increase capacity of the bag

        Precondition:   newCapacity must be an integer and > 0

        Postcondition:  The bag's capacity is at least newCapacity.
                        If the capacity was already at or greater than
                        newCapacity, then the capacity is left
                        unchanged.

        Args:           newCapacity (int)

        Raises:         TypeError if newCapacity is not an integer
                        ValueError if newCapacity not > 1
                        MemoryError if dynamic memory exhausted
        '''
        if not isinstance(newCapacity, int):
            raise TypeError("newCapacity must be an integer")
        if not newCapacity >= 1:
            raise ValueError("newCapacity must be >= 1")

        if len(self.__data) < newCapacity:
            # STUDENT WORK GOES HERE
            #   create a new, bigger array with the capacity
            #   newCapacity, copy all the elements from
            #   self.__data into the new array, and then
            #   update the reference self.__data to refer
            #   to the new array

            new_data = array.array('d', range(0, newCapacity))

            for i in range(self.__used):
                new_data[i] = self.__data[i]
            # end for

            self.__data = new_data
        # end if
    # end ensureCapacity

    def getCapacity(self) -> int:
        '''
        Returns:  an integer representing capacity of the bag.
        '''
        return len(self.__data)
    # end getCap acity

    def trimToSize(self) -> None:
        '''
        reduce capacity to current size if there is excess capacity

        Postcondition:  Capacity is reduced to current number of items
                        in bag or left unchanged if capacity equals to
                        number of items in bag

        Raises:         MemoryError if not enough dynamic memory to
                        support additional allocation

        '''
        if self.__used < len(self.__data):
            if self.__used <= 1:
                newCapacity = 1
            else:
                newCapacity = self._used
            # end if
            # STUDENT WORK HERE
            #   must create a new smaller array with capacity
            #   newCapacity, copy all the items from self.__data
            #   into the new array, and then update the reference
            #   self.__data to refer to the new array

            new_data = array.array('d', range(0, newCapacity))

            for i in range(self.__used):
                new_data[i] = self.__data[i]
            # end for

            self.__data = new_data
        # end if
    #  end trimToSize

    def insert(self, newEntry: Number) -> None:
        '''
        adds newEntry to the DoubleArrayBag if capacity available;

        Precondition:  newEntry must be a double; otherwise,
                       method raises a TypeError exception.

        Postcondition: newEntry is placed in Bag just after
                       items already existing and self.__used
                       is incremented by 1.

        Args:          newEntry (float)

        Raises:        MemoryError if not enough memory to make insertion

        '''
        if not (isinstance(newEntry, float) or (isinstance(newEntry, int))):
                raise TypeError("newEntry must be a double")
        # end if

        if  self.__used >= len(self.__data):
        # STUDENT WORK HERE
        #   must ensure that we have the capacity to hold the
        #   newly inserted item -- if not double the current
        #   capacity; then do a normal insertion

            self.ensureCapacity(len(self.__data) * 2)
        # end if

        self.__data[self.__used] = newEntry
        self.__used += 1
    # end add

    def __iadd__(self, addend: DoubleArrayBag) -> DoubleArrayBag:
        '''
        an implementation of the += operation

        Precondition:   addend is an instance of DoubleArrayBag and is not None

        Postcondition:  the elements from addend have been added to this bag

        Raises:         MemoryError if not enough memory
                        available to accommodate operation

                        TypeError if added is None or is not an instance of
                        DoubleArrayBag
        '''
        if (addend is None) or not(isinstance(addend,  self.__class__)):
            raise TypeError("addend is None or not instance of DoubleArrayBag")
        # end if
        # STUDENT WORK HERE
        #   guarantee that we have exactly enough room to hold all the items
        #   from addend and then copy them into this bag; then update any
        #   attributes to ensure that we are maintaining the class
        #   invariant

        if (addend.__used > 0):
            self.ensureCapacity(self.__used + addend.__used)

            amt = addend.__used
            for i in range(amt):
                self.__data[self.__used] = addend.__data[i]
                self.__used += 1
            # end for
        # end if

        return self
    # end plusEqual

    def occurrences(self, target: Number) -> int:
        '''
        counts occurrences of target in the bag

        Precondition:   target must be a float; otherwise,
                        method raises a TypeError exception

        Postcondition:  bag's state is not changed

        Arg:            target (float)

        Returns:        an integer number of occurrences of
                        target in the bag
        '''
        if not (isinstance(target, float) or (isinstance(target, int))):
            raise TypeError("target must be a float")
        count = 0
        for index in range(0, self.__used):
            if self.__data[index] == target:
                count += 1
            # end if
        # end for
        #
        # Below statement is the "magic" and lazy way to count occurrences
        # return self.__data.count(target)
        return count
    # end occurrences

    def erase_one(self, target: Number) -> bool:
        '''
        removes one copy of target from the DoubleArrayBag if target exists
        in the bag

        Precondition:  target must be a float; otherwise,
                       method raises a TypeError exception.

        Postcondition: if target exists in the bag, the first
                       occurrence of target is removed and
                       used is decremented.  Otherwise,
                       the bag's state is the same as it was
                       before the call.

        Arg:           target which must be an int

        Returns:       False or True depending on whether
                       target can be removed.
        '''
        if not (isinstance(target, float) or (isinstance(target, int))):
            raise TypeError("target must be a float")
        # end if

        found = False
        index = 0
        while not found and index < self.__used:
            if target == self.__data[index]:
                found = True
            else:
                index += 1
            # end if
        # end while
        if found:
            # we know we found target at data[index]
            # now reduce used by 1 and copy the last element
            # onto data[index] since ordering doesn't matter
            self.__used -= 1
            self.__data[index] = self.__data[self.__used]
        # end if
        return found
    # end erase_one

    def erase(self, target: Number) -> int:
        '''
        removes all copies of target from the DoubleArrayBag if target
        exists in the bag

        Precondition:  target must be an integer; otherwise,
                       method raises a TypeError exception.

        Postcondition: if target exists in the bag, all
                       occurrences of target are removed and
                       used is decremented by number of
                       items removed.  Otherwise, the bag's state
                       is the same as it was before the call.

        Arg:           target which must be an int

        Returns:       an int representing number of items erased
        '''
        if not (isinstance(target, float) or (isinstance(target, int))):
            raise TypeError("target must be a float")
        # end if

        number_removed = 0
        index = 0
        while index < self.__used:
            if self.__data[index] == target:
                self.__used -= 1
                self.__data[index] = self.__data[self.__used]
                number_removed += 1
            else:
                index += 1
            # end if
        # end while

        return number_removed
    # end erase

    @staticmethod
    def union(b1, b2) -> DoubleArrayBag:
        '''
        this method combines all the elements in this bag and all the
        elements in the other bag by placing them in a new bag

        Also note that this is a class method NOT an instance method

        Precondition:       b1 and b2 must be an instance of DoubleArrayBag
                            otherwise TypeError is raised

        Postcondition:      a new DoubleArrayBag is created containing copies
                            of all the elements in b1 plus all the elements
                            in b2.

        Args:               b1 and b2 (DoubleArrayBag)

        Raises:             TypeError if b1 or b2 not an instance of
                            DoubleArrayBag

                            MemoryError if note enough dynamic memory
                            to handle the union

        Returns:            A new DoubleArray bag that is the logical union
                            of the two arguments
        '''

        if not(isinstance(b1, DoubleArrayBag)):
            raise TypeError("b1 not instance of DoubleArrayBag")
        # end if
        if not(isinstance(b2, DoubleArrayBag)):
            raise TypeError("b2 not instance of DoubleArrayBag")
        # end if

        # STUDENT WORK HERE
        #   We must create (and ultimately return) a new bag that has exactly
        #   the capacity needed to hold all items from both b1 and b2 and
        #   then copy the items from b1 and then b2 into the new bag -- caution,
        #   watch out when both bags are empty -- be sure to test this

        new_bag = b1.copy()
        new_bag += b2

        return new_bag
    # end union

    def __str__(self) -> str:
        '''
        renders a string representation of the bag

        Postcondition:  bag remains in state it was in prior to call

        Returns:        a string representing current state of the bag
        '''
        result = "DoubleArrayBag with "
        result += str(self.__used)
        result += " elements: ["
        if self.__used > 0:
            result += str(self.__data[0])
            for index in range(1, self.__used):
                result += ", " + str(self.__data[index])
            # end for
        # end if
        result += " ]"
        result += " Capacity: " + str(self.getCapacity())

        return result
    # end __str__

    def __repr__(self) -> str:
        return str(self)
    # end __repr__

# end class Bag


def main():

    ''' below testing is NOT formal testing -- rather it is ad hoc '''
    ages = DoubleArrayBag()
    print("%s %s" % ("After constructor: ", ages))

    ages = DoubleArrayBag()
    print("%s %s" % ("After constructor: ", ages))
    ages.insert(47)
    ages.insert(33)
    ages.insert(17)
    ages.insert(2147483647)
    ages.insert(-2147483648)
    ages.insert(99)
    print("%s\n%s" %
          ("After inserting 47, 33, 17, 2147483647, 2147483648 and 99", ages))
    ages.erase(17)
    print("%s\n%s" % ("After attempting to remove 17", ages))
    ages.erase(99)
    print("%s\n%s" % ("After attempting to remove 99", ages))
    ages.erase(2147483647)
    print("%s\n%s" % ("After attempting to remove 2147483647", ages))
    ages.erase(47)
    print("%s\n%s" % ("After attempting to remove 47", ages))

    a = ages.copy()
    print("%s\n%s" % ("a after being copied from ages", a))
    print("%s %s" % (" a == ages ", str(a == ages)))
    print("%s %s" % (" not a is ages ", str(a is not ages)))

    b = DoubleArrayBag.union(ages, a)
    print("%s\n%s" % ("b after b = DoubleArrayBag.union( ages, a )", b))
# end main


if __name__ == "__main__":
    main()
# end if
